// lwc-services/lib/commands/sniff.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const fs = require("fs");
const path = require("path");
const util = require("util");
const webpackMerge = require("webpack-merge");
const jestConfig_1 = require("../config/jestConfig");
const lwcConfig_1 = require("../config/lwcConfig");
const webpack_config_1 = require("../config/webpack.config");
const sniff_1 = require("../messages/sniff");
const logger_1 = require("../utils/logger");
class Sniff extends command_1.Command {
    async run() {
        const { flags } = this.parse(Sniff);
        // eslint-disable-next-line no-console
        console.clear();
        logger_1.welcome();
        if (!flags.directory) {
            logger_1.log(sniff_1.messages.errors.nodirectory);
            return;
        }
        if (!fs.existsSync(flags.directory)) {
            fs.mkdirSync(flags.directory);
        }
        logger_1.log(sniff_1.messages.logs.calculating_configurations);
        let webpackConfig = webpack_config_1.generateWebpackConfig();
        if (flags.webpack) {
            // Merging custom webpack config file
            logger_1.log(sniff_1.messages.logs.custom_configuration);
            const webpackConfigCustom = require(path.resolve(process.cwd(), flags.webpack));
            webpackConfig = webpackMerge.smart(webpackConfig, webpackConfigCustom);
        }
        logger_1.log(sniff_1.messages.logs.write_jest_config);
        const inspectOptions = { depth: null };
        fs.writeFileSync(path.join(flags.directory, 'jest.config.js'), 'module.exports = ' + util.inspect(jestConfig_1.jestConfig, inspectOptions));
        logger_1.log(sniff_1.messages.logs.write_webpack_config);
        fs.writeFileSync(path.join(flags.directory, 'webpack.config.js'), 'module.exports = ' + util.inspect(webpackConfig, inspectOptions));
        logger_1.log(sniff_1.messages.logs.write_lwc_config);
        fs.writeFileSync(path.join(flags.directory, 'lwc-services.config.js'), 'module.exports = ' + util.inspect(lwcConfig_1.defaultLwcConfig, inspectOptions));
        logger_1.log(sniff_1.messages.logs.enjoy);
    }
}
exports.default = Sniff;
Sniff.description = sniff_1.messages.description;
Sniff.examples = sniff_1.messages.help.examples;
Sniff.flags = {
    help: command_1.flags.help({ char: 'h' }),
    directory: command_1.flags.string({
        char: 'd',
        description: sniff_1.messages.flags.directory,
        required: true
    }),
    webpack: command_1.flags.string({
        char: 'w',
        description: sniff_1.messages.flags.webpack
    })
};
