// proxy-compat/build/intrinsics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
var _a = Object, getOwnPropertyNames = _a.getOwnPropertyNames, create = _a.create, keys = _a.keys, getOwnPropertyDescriptor = _a.getOwnPropertyDescriptor, preventExtensions = _a.preventExtensions, defineProperty = _a.defineProperty, hasOwnProperty = _a.hasOwnProperty, isExtensible = _a.isExtensible, getPrototypeOf = _a.getPrototypeOf, setPrototypeOf = _a.setPrototypeOf;
exports.getOwnPropertyNames = getOwnPropertyNames;
exports.create = create;
exports.keys = keys;
exports.getOwnPropertyDescriptor = getOwnPropertyDescriptor;
exports.preventExtensions = preventExtensions;
exports.defineProperty = defineProperty;
exports.hasOwnProperty = hasOwnProperty;
exports.isExtensible = isExtensible;
exports.getPrototypeOf = getPrototypeOf;
exports.setPrototypeOf = setPrototypeOf;
var _b = Array.prototype, ArraySlice = _b.slice, ArrayShift = _b.shift, ArrayUnshift = _b.unshift, ArrayConcat = _b.concat;
exports.ArraySlice = ArraySlice;
exports.ArrayShift = ArrayShift;
exports.ArrayUnshift = ArrayUnshift;
exports.ArrayConcat = ArrayConcat;
var isArray = Array.isArray;
exports.isArray = isArray;
