// es5-proxy-compat/module.js
/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
// Spaces in this file indicate helpers that we don't yet support
// Order is alphabetical
import applyDecoratedDescriptor from 'babel-compat/helpers/applyDecoratedDescriptor';
import arrayWithHoles from 'babel-compat/helpers/arrayWithHoles';
import arrayWithoutHoles from 'babel-compat/helpers/arrayWithoutHoles';
import assertThisInitialized from 'babel-compat/helpers/assertThisInitialized';

import asyncGeneratorDelegate from 'babel-compat/helpers/asyncGeneratorDelegate';
import asyncIterator from 'babel-compat/helpers/asyncIterator';
import asyncToGenerator from 'babel-compat/helpers/asyncToGenerator';
import awaitAsyncGenerator from 'babel-compat/helpers/awaitAsyncGenerator';

import classCallCheck from 'babel-compat/helpers/classCallCheck';
import classNameTDZError from 'babel-compat/helpers/classNameTDZError';


import _constructClass from 'babel-compat/helpers/construct';
import createClass from 'babel-compat/helpers/createClass';

import defaults from 'babel-compat/helpers/defaults';
import defineEnumerableProperties from 'babel-compat/helpers/defineEnumerableProperties';
import _defineProperty from 'babel-compat/helpers/defineProperty';
import _extends from 'babel-compat/helpers/extends';
import _get from 'babel-compat/helpers/get';
import _getPrototypeOf from 'babel-compat/helpers/getPrototypeOf';
import inherits from 'babel-compat/helpers/inherits';
import inheritsLoose from 'babel-compat/helpers/inheritsLoose';
import initializerDefineProperty from 'babel-compat/helpers/initializerDefineProperty';
import initializerWarningHelper from 'babel-compat/helpers/initializerWarningHelper';
import _instanceof from 'babel-compat/helpers/instanceof';
import isNativeFunction from 'babel-compat/helpers/isNativeFunction';
import iterableToArray from 'babel-compat/helpers/iterableToArray';
import iterableToArrayLimit from 'babel-compat/helpers/iterableToArrayLimit';
import iterableToArrayLimitLoose from 'babel-compat/helpers/iterableToArrayLimitLoose';

import newArrowCheck from 'babel-compat/helpers/newArrowCheck';
import nonIterableRest from 'babel-compat/helpers/nonIterableRest';
import nonIterableSpread from 'babel-compat/helpers/nonIterableSpread';
import objectDestructuringEmpty from 'babel-compat/helpers/objectDestructuringEmpty';
import objectSpread from 'babel-compat/helpers/objectSpread';
import objectWithoutProperties from 'babel-compat/helpers/objectWithoutProperties';

import possibleConstructorReturn from 'babel-compat/helpers/possibleConstructorReturn';
import readOnlyError from 'babel-compat/helpers/readOnlyError';
import _set from 'babel-compat/helpers/set';
import _setPrototypeOf from 'babel-compat/helpers/setPrototypeOf';
import skipFirstGeneratorNext from 'babel-compat/helpers/skipFirstGeneratorNext';
import slicedToArray from 'babel-compat/helpers/slicedToArray';

import superPropBase from 'babel-compat/helpers/superPropBase';
import taggedTemplateLiteral from 'babel-compat/helpers/taggedTemplateLiteral';

import temporalRef from 'babel-compat/helpers/temporalRef';
import _temporalUndefined from 'babel-compat/helpers/temporalUndefined';
import toArray from 'babel-compat/helpers/toArray';
import toConsumableArray from 'babel-compat/helpers/toConsumableArray';
import toPropertyKey from 'babel-compat/helpers/toPropertyKey';
import _typeof from 'babel-compat/helpers/typeof';
import wrapAsyncGenerator from 'babel-compat/helpers/wrapAsyncGenerator';
import _wrapNativeSuper from 'babel-compat/helpers/wrapNativeSuper';

import regenerator from 'babel-compat/regenerator';

const babelHelpers = {
    applyDecoratedDescriptor,
    assertThisInitialized,
    arrayWithHoles,
    arrayWithoutHoles,
    asyncGeneratorDelegate,
    asyncIterator,
    asyncToGenerator,
    awaitAsyncGenerator,
    classCallCheck,
    classNameTDZError,
    construct: _constructClass,
    createClass,
    defaults,
    defineEnumerableProperties,
    defineProperty: _defineProperty,
    extends: _extends,
    get: _get,
    getPrototypeOf: _getPrototypeOf,
    inherits,
    inheritsLoose,
    initializerDefineProperty,
    initializerWarningHelper,
    iterableToArray,
    iterableToArrayLimit,
    iterableToArrayLimitLoose,
    instanceof: _instanceof,
    isNativeFunction,
    newArrowCheck,
    nonIterableRest,
    nonIterableSpread,
    objectDestructuringEmpty,
    objectSpread,
    objectWithoutProperties,
    possibleConstructorReturn,
    readOnlyError,
    set: _set,
    setPrototypeOf: _setPrototypeOf,
    skipFirstGeneratorNext,
    slicedToArray,
    superPropBase,
    taggedTemplateLiteral,
    temporalRef,
    toArray,
    toConsumableArray,
    toPropertyKey,
    typeof: _typeof,
    temporalUndefined: _temporalUndefined,
    wrapNativeSuper: _wrapNativeSuper,
    wrapAsyncGenerator,
};

export {
    regenerator,
    babelHelpers,
};
