// lwc-services/lib/config/lwcConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const merge = require("deepmerge");
const path = require("path");
exports.defaultLwcConfig = {
    // Default directory for the build output
    buildDir: './dist',
    // Default mode for build command
    mode: 'development',
    // Clears the build directory on every build
    noclear: false,
    // Defines the directory where to find the individual
    // modules. Change only when you know what you do. ;-)
    moduleDir: 'src/modules',
    // Array of directories where to look for additional
    // modules that don't live in `moduleDir`
    localModulesDirs: [],
    // Defines the directory layout. Using `namespaced` is easiest. Or so.
    layout: 'namespaced',
    // Default directory for source files
    sourceDir: './src',
    // List of resources for copying to the build folder
    resources: [],
    // Default server options for watch command
    devServer: {
        port: 3001,
        host: '0.0.0.0',
        open: false,
        stats: 'errors-only',
        noInfo: true,
        contentBase: './src'
    },
    // Default server options for serve command
    server: {
        port: 3002,
        host: '0.0.0.0',
        open: false
    },
    // LWC Compiler options for production mode.
    // Find the detailed description here: https://www.npmjs.com/package/@lwc/compiler
    lwcCompilerOutput: {
        development: {
            compat: false,
            minify: true,
            env: {
                NODE_ENV: 'production'
            },
            format: 'amd'
        },
        production: {
            compat: false,
            minify: true,
            env: {
                NODE_ENV: 'production'
            },
            format: 'amd'
        }
    },
    lwcCompilerStylesheetConfig: {},
    lwcExperimentalDynamicComponent: {}
};
function buildConfig() {
    let combinedConfig = exports.defaultLwcConfig;
    try {
        const fileName = path.resolve(process.cwd(), 'lwc-services.config.js');
        const config = require(fileName);
        combinedConfig = merge(exports.defaultLwcConfig, config);
        return combinedConfig;
    }
    catch (error) {
        // We don't need error handling atm
    }
    return combinedConfig;
}
exports.lwcConfig = buildConfig();
